V1 attempt to solve GH Action error

willdavidson05 opened the pull request:
<!-- _modified from [EmbeddedArtistry](https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/)_
_referenced with modifications from [pycytominer](https://github.com/cytomining/pycytominer/blob/master/.github/PULL_REQUEST_TEMPLATE.md)_ -->

# Description

<!--
Thank you for your contribution to the Software Gardening Almanack!

Please _succinctly_ summarize your proposed change.
Namely, consider addressing the following questions:

- What motivated you to open this pull request?
- Were there any special adjustments you had to make to complete the work?
- Are there any issues which are related to this pull request (you may used a `#<digit>` to reference GitHub issues as links within this description)?

Also, if you haven't already, please use `pre-commit run --all-files` to help check your files using this project's pre-commit configuration.
-->

## What is the nature of your change?

- [ ] Content additions or updates (adds or updates content)
- [ ] Bug fix (fixes an issue).
- [ ] Enhancement (adds functionality).
- [ ] Breaking change (these changes would cause existing functionality to not work as expected).

# Checklist

Please ensure that all boxes are checked before indicating that this pull request is ready for review.

- [ ] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own contributions.
- [ ] I have commented my content, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to related documentation (outside of book content).
- [ ] My changes generate no new warnings.
- [ ] New and existing tests pass locally with my changes.
- [ ] I have added tests that prove my additions are effective or that my feature works.
- [ ] I have deleted all non-relevant text in this pull request template.

d33bs commented on .github/workflows/entropy-check.yml:
```diff
@@ -22,24 +22,24 @@ jobs:
         run: |
-          "echo 'PR_BRANCH=${PR_BRANCH}' >> $GITHUB_ENV"
-          "echo 'MAIN_BRANCH=${MAIN_BRANCH}' >> $GITHUB_ENV"
+          export PR_BRANCH=${{ github.head_ref }}
+          export MAIN_BRANCH=${{ github.base_ref }}
+          echo "PR_BRANCH is $PR_BRANCH"
+          echo "MAIN_BRANCH is $MAIN_BRANCH"
```
When running these commands together like this you typically need to use `&&` or `||` to help them chain. [See here](https://unix.stackexchange.com/a/24685) for more information on how these work. I think the current version of this attempts to run something like: `export PR_BRANCH=${{ github.head_ref }} export MAIN_BRANCH=${{ github.base_ref }} echo "PR_BRANCH is $PR_BRANCH" echo "MAIN_BRANCH is $MAIN_BRANCH"` which I believe will result in errors (the first export command will interpret everything else as flags or arguments). The newlines might also get added in here due to the `|` string continuation within yaml. I think the GitHub Actions job is maybe truncating the output, which makes it a bit confusing to look at.

Consider the following:

You can always view the exit status code of previous commands with: `echo $?`. Each command generally should exit with a zero, which means no errors or warnings occurred (be careful here, sometimes the process might "want" to find errors successfully and exit with a 0 because it did it's job successfully). If it doesn't exit with zero we generally intercept that as an error. We can also simplify this with `True` and `False` in bash to simulate what happens to commands.

```bash
# 1 "and" 0
False && True
echo $?
# shows 1

# 0 "and" 0
True && True
echo $?
# shows 0

# 1 "or" 0
False || True
echo $?
# shows 0

# 0 "or" 0
True || True
echo $?
# shows 0
```

GitHub Actions I think will check the status of the steps and because it sees the `export` didn't finish successfully (with a 0) it gathers this up for us in the job and tells us something went wrong.

d33bs replied:
Looking at this again, I think I'm wrong here about the need for the `&&`! Leaving this up as there's still some useful information here just in case. I'm going to ponder this a bit more, apologies.

d33bs commented on .github/workflows/entropy-check.yml:
````diff
@@ -22,24 +22,24 @@ jobs:
-          --repo_path '.' \
-          --pr_branch '${{ env.PR_BRANCH }}' \
-          --main_branch '${{ env.MAIN_BRANCH }}' > entropy_report.txt"
+          ```poetry run almanack process_pr_entropy \
````
I'm not certain but I believe the surrounding backticks here may cause issues.

d33bs commented on .github/workflows/entropy-check.yml:
```diff
@@ -22,24 +22,24 @@ jobs:
         run: |
-          "echo 'PR_BRANCH=${PR_BRANCH}' >> $GITHUB_ENV"
-          "echo 'MAIN_BRANCH=${MAIN_BRANCH}' >> $GITHUB_ENV"
+          export PR_BRANCH=${{ github.head_ref }}
```
Since these are github values you're referencing, would it be possible for you to reference them directly instead of re-setting them as environment values?

d33bs reviewed (commented):
Hi @willdavidson05 - I took a deeper look at this an realized what might be happening. It looks like the inclusion of double and single quotes for the environment variables is messing with things. The double quotes wrapped around the whole line will attempt to run a command which is named as the double quoted string. In addition, I believe you may want to wrap the environment variables within double quotes, just in case the single quotes cause any sort of issue. See here for more: https://stackoverflow.com/questions/6697753/difference-between-single-and-double-quotes-in-bash


Here's an example of what things look like:
- source: https://github.com/d33bs/demo-gh-actions-env-vars/blob/main/.github/workflows/demo-env-vars.yml
- processed output: https://github.com/d33bs/demo-gh-actions-env-vars/actions/runs/11055753786/job/30715662572

Based on this, would you like to open a PR on the upstream with a fix once you feel comfortable with things?

willdavidson05 commented:
@d33bs Thank you for all the help on this! It makes sense that the mixture of quotes would be causing the problem, especially because the difference between the two was not clear to me, haha. I appreciate the links/resources and help on this issue. I definitely open an upstream PR with a fix once I implement your feedback! Thanks again!!

willdavidson05 commented:
@d33bs I added the double quotes and removed all single quotes even down to the formatting echo, but I am still getting the same error. I've messed around with debugging steps such as ensuring that GITHUB_ENV exists before writing, but it hasn't helped me identify the cause yet. I've also tried looking into the logs with the GitHub CLI, but using this I was also not able to make any progress. 

I have added two new echos that set the two different branch environments (attempting to modulate the problem):
```        # Set PR Branch Environment Variable (PR branch)
        - name: Set PR Branch Environment Variable
          run: echo "PR_BRANCH=${{ github.head_ref }}" >> "$GITHUB_ENV"

        # Set Main Branch Environment Variable (main branch)
        - name: Set Main Branch Environment Variable
          run: echo "MAIN_BRANCH=${{ github.base_ref }}" >> "$GITHUB_ENV"  
```

And also simplified the entropy report format: 

```    - name: Save Entropy Report
          run: |
            echo "REPORT_CONTENT<<EOF" >> "$GITHUB_ENV"
            cat entropy_report.txt >> "$GITHUB_ENV"
            echo "EOF" >> "$GITHUB_ENV "
```

I'm not sure if I missed anything from your feedback but please let me know if you see anything that jumps out! Thanks again for the help with this very persistent error!
